packages/@lwc/integration-karma/test/rendering/callback-invocation-order/index.spec.js

```diff
@@ -297,3 +299,90 @@ it('should invoke callbacks on the right order when multiple templates are used
             ]);
         });
 });
+
+if (process.env.NATIVE_SHADOW) {
+    // #TODO[3827]: These regression tests can be removed once the bug is resolved.
+    it('regression test (#3827) - shadow DOM should maintain callback invocation order for duplicate slots across conditional branches when rerender triggered on conditional slot component', () => {
+        const container = createElement('x-shadow-container-multiple-conditionals', {
+            is: ShadowContainerMultipleConditionals,
+        });
+        document.body.appendChild(container);
+
+        const currentLeafName = container.getLeaf().name;
+        expect(window.timingBuffer).toEqual([
+            'shadowContainer:connectedCallback',
+            'shadowSlot:connectedCallback',
+            `leaf:${currentLeafName}:connectedCallback`,
+        ]);
+
+        resetTimingBuffer();
+
+        const slotCmp = container.getSlotCmp();
+        slotCmp.showIf = true;
+
+        return Promise.resolve()
+            .then(() => {
+                expect(window.timingBuffer).toEqual([]);
+
+                slotCmp.showIf = false;
+                slotCmp.showElseIf = true;
+            })
+            .then(() => {
+                expect(window.timingBuffer).toEqual([]);
+                document.body.removeChild(container);
+            });
+    });
+}
+
+it('regression test (#3827) - light DOM should maintain callback invocation order for duplicate slots across conditional branches when rerender triggered on conditional slot component', () => {
+    // #TODO[3827]: These regression tests can be removed once the bug is resolved.
+    const container = createElement('x-light-container-multiple-conditionals', {
+        is: LightContainerMultipleConditionals,
+    });
+    document.body.appendChild(container);
+
+    let currentLeafName = container.getLeaf().name;
+    expect(window.timingBuffer).toEqual([
+        'lightContainer:connectedCallback',
+        'lightSlot:connectedCallback',
+        `leaf:${currentLeafName}:connectedCallback`,
+    ]);
+
+    resetTimingBuffer();
+
+    const slotCmp = container.getSlotCmp();
+    slotCmp.showIf = true;
+
+    return Promise.resolve()
+        .then(() => {
+            const previousLeafName = currentLeafName;
+            currentLeafName = container.getLeaf().name;
+
+            expect(window.timingBuffer).toEqual(
+                window.lwcRuntimeFlags.ENABLE_NATIVE_CUSTOM_ELEMENT_LIFECYCLE
+                    ? [
+                          `leaf:${currentLeafName}:connectedCallback`,
+                          `leaf:${previousLeafName}:disconnectedCallback`,
+                          `leaf:${currentLeafName}:disconnectedCallback`,
+                      ]
+                    : [
+                          `leaf:${currentLeafName}:connectedCallback`,
+                          `leaf:${currentLeafName}:disconnectedCallback`,
+                      ]
```

Comment:
So the previous leaf never fires a disconnectedCallback? I thought this was the entire bug?

Reply:
@nolanlawson yup, the previous leaf won't fire `disconnectedCallback` in synthetic lifecycle.

This is how it behaves in 3.x.x, this PR just brings the behavior back to what it was in 3.x.x.

I plan to raise another PR to fix the `disconnectedCallback` behavior entirely as part of #3832

Reply:
> I plan to raise another PR to fix the `disconnectedCallback` behavior entirely

Is this worth doing if we are going to do native lifecycle in 5.0?

Reply:
Yeah I think either way this will need to be fixed because if you look at the native lifecycle (one above this one) the new component is connected and immediately disconnected.

Which means all the components are disconnected after `unmount` runs.

Reply:
Also, this is a very specific bug, the rerender has to be isolated to just the component containing the `<slot>` element.

For example:

```html

<template>
    <template lwc:if={showMe}>
        <slot></slot>
    <template lwc:else>
        <slot></slot>
</template>
```
If the value of `showMe` is passed down from a parent component then this bug does not appear, but if the value of `showMe` is isolated to just the component, it does.

Reply:
> Yeah I think either way this will need to be fixed because if you look at the native lifecycle (one above this one) the new component is connected and immediately disconnected.

So is the issue that the diffing algo is doing extra unnecessary work?

Reply:
Sort of, in this specific bug, the references to what's being slotted is getting mixed up, so when the element is removed from the DOM both native and synthetic lifecycles are called.